File: basics/23-remove-duplicates-ll/Solution.java
import java.util.*;

class ListNode {
    int val;
    ListNode next;
    ListNode(int val) { this.val = val; }
}

class Solution {
    public ListNode removeDuplicates(ListNode head) {
        if (head == null) return null;
        Set<Integer> seen = new HashSet<>();
        ListNode curr = head;
        seen.add(curr.val);
        while (curr.next != null) {
            if (seen.contains(curr.next.val)) {
                curr.next = curr.next.next;
            } else {
                seen.add(curr.next.val);
                curr = curr.next;
            }
        }
        return head;
    }
}

File: basics/23-remove-duplicates-ll/Solution.cpp
#include <unordered_set>

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};

ListNode* removeDuplicates(ListNode* head) {
    if (!head) return nullptr;
    std::unordered_set<int> seen;
    ListNode* curr = head;
    seen.insert(curr->val);
    while (curr->next) {
        if (seen.count(curr->next->val)) {
            curr->next = curr->next->next;
        } else {
            seen.insert(curr->next->val);
            curr = curr->next;
        }
    }
    return head;
}
